src/App.tsx

```diff
@@ -0,0 +1,800 @@
+    (event: MouseEvent): void => {
+      if (!dataset) {
+        return;
+      }
+      const id = canv.getIdAtPixel(event.offsetX, event.offsetY);
+      if (id === BACKGROUND_ID) {
+        // Ignore background pixels
+        return;
+      }
+      setHoveredId(id);
+    },
+    [dataset, canv]
+  );
+
+  const onMouseLeave = useCallback((_event: MouseEvent): void => {
+    setHoveredId(null);
+  }, []);
+
+  useEffect(() => {
+    canv.domElement.addEventListener("mousemove", onMouseMove);
+    canv.domElement.addEventListener("mouseleave", onMouseLeave);
+    return () => {
+      canv.domElement.removeEventListener("mousemove", onMouseMove);
+      canv.domElement.removeEventListener("mouseleave", onMouseLeave);
+    };
+  }, [onMouseMove, onMouseLeave, canv]);
+
+  // SCRUBBING CONTROLS ////////////////////////////////////////////////////
+
+  const handleKeyDown = useCallback(
+    ({ key }: KeyboardEvent): void => {
+      if (key === "ArrowLeft" || key === "Left") {
+        timeControls?.handleFrameAdvance(-1);
+      } else if (key === "ArrowRight" || key === "Right") {
+        timeControls?.handleFrameAdvance(1);
+      }
+    },
+    [timeControls]
+  );
+
+  useEffect(() => {
+    window.addEventListener("keydown", handleKeyDown);
+
+    return () => {
+      window.removeEventListener("keydown", handleKeyDown);
+    };
+  }, [handleKeyDown]);
+
+  // Store the current value of the time slider as its own state, and update
+  // the frame using a debounced value to prevent constant updates as it moves.
+  const debouncedFrameInput = useDebounce(frameInput, 250);
+  useEffect(() => {
+    setFrame(debouncedFrameInput);
+  }, [debouncedFrameInput]);
+
+  const handleFindTrack = useCallback(async (): Promise<void> => {
+    // Load track value
+    await findTrack(parseInt(findTrackInput, 10));
+  }, [findTrackInput, findTrack]);
+
+  // RECORDING CONTROLS ////////////////////////////////////////////////////
+
+  // Update the callback for TimeControls and RecordingControls if it changes.
+  // TODO: TimeControls and RecordingControls should be refactored into components
+  // and receive setFrame as props.
+  useMemo(() => {
+    timeControls?.setFrameCallback(setFrame);
+    recordingControls?.setFrameCallback(setFrame);
+  }, [setFrame]);
+
+  // Update the default recording controls prefix.
+  useMemo(() => {
+    if (useDefaultPrefix && datasetName && featureName) {
+      setImagePrefix(`${datasetName}-${featureName}-`);
+    }
+  }, [useDefaultPrefix, datasetName, featureName]);
+
+  // RENDERING /////////////////////////////////////////////////////////////
+
+  const disableUi: boolean = recordingControls?.isRecording() || !datasetOpen;
+  const disableTimeControlsUi = disableUi;
+
+  return (
+    <div>
+      {/** Top Control Bar */}
+      <div className={styles.canvasTopControlsContainer}>
+        <label htmlFor="dataset">Dataset</label>
+        <select
+          name="Dataset"
+          id="dataset"
```

Comment:
I'd prefer using `className` to `id` in React components. Converting current uses of `getElementById` to refs should eliminate the need for unique IDs.

Reply:
Fixed!